File: app/services/filtering.py
from __future__ import annotations


def filter_games(games: list[dict], query: dict) -> list[dict]:
    team = query.get("team")
    matchup = query.get("matchup")
    if team and team != "ALL":
        games = [
            game
            for game in games
            if game["awayTeam"]["abbreviation"] == team or game["homeTeam"]["abbreviation"] == team
        ]
    if matchup and matchup != "ALL":
        games = [game for game in games if game["matchupId"] == matchup]
    return games


def filter_hitters(hitters: list[dict], query: dict) -> list[dict]:
    team = query.get("team")
    matchup = query.get("matchup")
    handedness = query.get("handedness")
    if team and team != "ALL":
        hitters = [hitter for hitter in hitters if hitter["team"]["abbreviation"] == team]
    if matchup and matchup != "ALL":
        hitters = [hitter for hitter in hitters if hitter["matchupId"] == matchup]
    if handedness and handedness != "ALL":
        hitters = [hitter for hitter in hitters if hitter["bats"] == handedness]
    return hitters


def filter_pitchers(pitchers: list[dict], query: dict) -> list[dict]:
    team = query.get("team")
    matchup = query.get("matchup")
    if team and team != "ALL":
        pitchers = [pitcher for pitcher in pitchers if pitcher["team"]["abbreviation"] == team]
    if matchup and matchup != "ALL":
        pitchers = [pitcher for pitcher in pitchers if pitcher["matchupId"] == matchup]
    return pitchers

File: backend/src/services/filtering.ts
import type {
  AttackablePitcher,
  DailyAnalysisQuery,
  GameInfo,
  RankedHitter,
  RankedPitcher,
} from '@mlb-analyzer/shared';

const matchesTeam = (teamFilter: string | undefined, teamCode: string, opponentCode: string): boolean => {
  if (!teamFilter) {
    return true;
  }

  return teamCode === teamFilter || opponentCode === teamFilter;
};

const matchesHandedness = (filter: DailyAnalysisQuery['handedness'], handedness: string): boolean =>
  !filter || filter === 'ALL' || filter === handedness;

const matchesMatchup = (
  matchupFilter: string | undefined,
  matchupId: string,
  matchupLabel: string,
): boolean => !matchupFilter || matchupId === matchupFilter || matchupLabel === matchupFilter;

export const filterGames = (games: GameInfo[], query: DailyAnalysisQuery): GameInfo[] =>
  games.filter(
    (game) =>
      matchesTeam(query.team, game.homeTeam.abbreviation, game.awayTeam.abbreviation) &&
      matchesMatchup(query.matchup, game.matchupId, game.matchupLabel),
  );

export const filterHitters = (hitters: RankedHitter[], query: DailyAnalysisQuery): RankedHitter[] =>
  hitters.filter(
    (hitter) =>
      matchesTeam(query.team, hitter.team.abbreviation, hitter.opponent.abbreviation) &&
      matchesMatchup(query.matchup, hitter.matchupId, hitter.matchupLabel) &&
      matchesHandedness(query.handedness, hitter.bats),
  );

export const filterPitchers = <
  T extends RankedPitcher | AttackablePitcher,
>(
  pitchers: T[],
  query: DailyAnalysisQuery,
): T[] =>
  pitchers.filter(
    (pitcher) =>
      matchesTeam(query.team, pitcher.team.abbreviation, pitcher.opponent.abbreviation) &&
      matchesMatchup(query.matchup, pitcher.matchupId, pitcher.matchupLabel) &&
      matchesHandedness(query.handedness, pitcher.throwingHand),
  );
